Documentation on identifying objects after the removal of slugs (#4372)
nautobot/nautobot: nautobot/docs/development/apps/api/platform-features/uniquely-identify-objects.md

- **Usability:** Natural keys are more user-friendly and easier to remember.
- **Portability:** Natural keys can be used to identify an object in multiple contexts, such as in external applications or in different databases.

However, there are also some disadvantages to using natural keys:

- **Complexity:** Using natural keys can make it more difficult to change the underlying data structure without breaking existing integrations.
- **Performance:** Natural keys are not guaranteed to be indexed which can make lookups slower. Also, natural keys that use a combination of fields will require database joins to retrieve the object from the database.

In general, the decision of whether to use a primary key or a natural key depends on the specific requirements of the application. If uniqueness is critical, or if performance is a concern, a primary key may be the best choice. If human-readability is important, or if portability is a concern, a natural key may be the best choice.

### Using Primary Keys

> In Nautobot v2.0, all object view URLs use the primary key of the object. In the web UI and REST API, the object can be accessed by building a URL using the primary key. For example, to retrieve the device with primary key `00000000-0000-0000-0000-000000000000`, the REST API URL would be `/api/dcim/devices/00000000-0000-0000-0000-000000000000/`.

Reviewer: Nit: Should we use an actual uuid here instead of a string of zeroes?

Author: This is a valid uuid and I feel that a random string of characters is more distracting than all zeroes

> In the Python ORM, the object can be retrieved using the `get()` method of the model manager. For example, to retrieve the device with primary key `00000000-0000-0000-0000-000000000000`, the Python ORM call would be `Device.objects.get(pk='00000000-0000-0000-0000-000000000000')`. Once an instance of the object is retrieved, the primary key can be accessed using the `pk` attribute.

Reviewer: Nit: Should we use an actual uuid here instead of a string of zeroes?

> # Uniquely Identifying a Nautobot Object

Reviewer: I don't know where to put it but should we mention that `GitRepository` still has its `slug` field?

Author: I think that's out of scope for this document. It belongs in the 2.0 upgrade guide

## Primary Keys vs. Natural Keys

When designing an application for Nautobot, one of the key decisions is whether to use a primary key or a natural key to uniquely identify individual model instances. Here are some trade-offs to consider:

### Primary Keys

The primary key is a concrete database field and is the primary method for referencing an object or relating multiple objects in the Nautobot database. Here are some advantages of using primary keys:

- **Uniqueness:** Primary keys are guaranteed to be unique.
- **Performance:** Primary keys are always indexed in Nautobot, which makes lookups very fast.

However, there are also some disadvantages to using primary keys:

- **Usability**: Primary keys in Nautobot use UUIDs and can be difficult to work with directly.
- **Portability:** Primary keys are specific to a single database and cannot be used to identify an object in multiple contexts, such as in external applications or in different databases.

> ### Natural Keys

Reviewer: Should we throw a doc reference to `docs/development/core/natural-keys.md` somewhere here?

> A natural key is an identifier that is based on the natural attributes of a record, such as a platforms's name or in the case of some Nautobot models, a combination of fields. For example, the Prefix model's natural key is formed using a combination of the `prefix` field and the associated Namespace's `name`. Here are some advantages of using natural keys:

Reviewer 2: ```suggestion
A natural key is an identifier that is based on the natural attributes of a record, such as a platform's name or, in the case of some Nautobot models, a combination of fields. For example, the Prefix model's natural key is formed using a combination of the `prefix` field and the associated Namespace's `name`. Here are some advantages of using natural keys:
```

- **Usability:** Natural keys are more user-friendly and easier to remember.
- **Portability:** Natural keys can be used to identify an object in multiple contexts, such as in external applications or in different databases.

However, there are also some disadvantages to using natural keys:

- **Complexity:** Using natural keys can make it more difficult to change the underlying data structure without breaking existing integrations.

> - **Performance:** Natural keys are not guaranteed to be indexed which can make lookups slower. Also, natural keys that use a combination of fields will require database joins to retrieve the object from the database.

Reviewer 2: ```suggestion
- **Performance:** Natural keys are not guaranteed to be indexed which can make lookups slower. Also, natural keys that include fields from related objects will require database joins to retrieve the object from the database.
```
